File: ruby/orders/orders_workflow.rb
#!/usr/bin/env ruby
# Encoding: utf-8
#
# Copyright:: Copyright 2016, Google Inc. All Rights Reserved.
#
# License:: Licensed under the Apache License, Version 2.0 (the "License");
#           you may not use this file except in compliance with the License.
#           You may obtain a copy of the License at
#
#           http://www.apache.org/licenses/LICENSE-2.0
#
#           Unless required by applicable law or agreed to in writing, software
#           distributed under the License is distributed on an "AS IS" BASIS,
#           WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
#           implied.
#           See the License for the specific language governing permissions and
#           limitations under the License.
#
# Performs a simulated workflow for a single order on the sandbox API
# endpoint. We do this on the sandbox API endpoint for two reasons:
# 1) to have access to methods needed to create/advance test orders, and
# 2) to avoid accidental mutation of existing real orders.

require_relative 'orders_common'

def order_workflow(content_api, merchant_id)
  # Create a new test order using the template1 template. Normally orders would
  # be automatically populated by Google in the non-sandbox version, and we'd
  # skip ahead to find out what orders are currently waiting on us.
  print "Creating test order... "
  req = Google::Apis::ContentV2::OrdersCreateTestOrderRequest.new
  req.template_name = "template1"
  # For most of the calls in this workflow, we'll not check the return value
  # and instead use the simplified flow that just raises the error if it
  # happens and returns the response otherwise.
  resp = content_api.create_test_order(merchant_id, req)
  order_id = resp.order_id
  puts "done with id #{order_id}."
  puts

  # List all unacknowledged orders (a call like this is where we'd normally get
  # order_id values to operate on).
  puts "Listing unacknowledged orders for merchant #{merchant_id}:"
  list_all_orders(content_api, merchant_id, acknowledged: false)
  puts

  # Acknowledge the newly received order.
  print "Acknowledging order #{order_id}... "
  req = Google::Apis::ContentV2::OrdersAcknowledgeRequest.new
  req.operation_id = new_operation_id()
  resp = content_api.acknowledge_order(merchant_id, order_id, req)
  puts "done (#{resp.execution_status})."
  puts

  # Set the new order's merchant order ID.  We'll just use a random 32-bit int.
  merchant_order_id = "test order #{Random.rand(2**32 - 1)}"
  print "Updating merchant order ID to '#{merchant_order_id}'... "
  req = Google::Apis::ContentV2::OrdersUpdateMerchantOrderIdRequest.new
  req.merchant_order_id = merchant_order_id
  req.operation_id = new_operation_id()
  resp = content_api.update_merchant_order_id(merchant_id, order_id, req)
  puts "done (#{resp.execution_status})."
  puts

  print "Retrieving order with merchant order ID '#{merchant_order_id}'... "
  resp =
     content_api.get_order_by_merchant_order_id(merchant_id, merchant_order_id)
  print "done.\n\n"
  current_order = resp.order
  print_order(current_order)
  puts

  # Oops, not enough stock for all the Chromecasts ordered, so we cancel one
  # of them.
  print "Canceling one Chromecast order... "
  req = Google::Apis::ContentV2::OrdersCancelLineItemRequest.new
  req.operation_id = new_operation_id()
  req.line_item_id = current_order.line_items[0].id
  req.quantity = 1
  req.reason = "noInventory"
  req.reason_text = "Ran out of inventory while fulfilling request."
  resp = content_api.cancel_order_line_item(merchant_id, order_id, req)
  puts "done (#{resp.execution_status})."
  puts

  current_order = content_api.get_order(merchant_id, order_id)
  print_order(current_order)
  puts

  # Advance the test order to the shippable state. Normally this would be done
  # by Google when an order is no longer cancelable by the customer, but here
  # we need to do it manually.
  print "Advancing test order... "
  content_api.advance_test_order(merchant_id, order_id)
  puts "done."
  puts

  current_order = content_api.get_order(merchant_id, order_id)
  print_order(current_order)
  puts

  # To simulate partial fulfillment, we'll pick the first line item and ship
  # the still-pending amount.
  print "Notifying Google about shipment of first line item... "
  line_item = Google::Apis::ContentV2::OrderShipmentLineItemShipment.new
  line_item.line_item_id = current_order.line_items[0].id
  line_item.quantity = current_order.line_items[0].quantity_pending
  # Storing this in a non-reused variable so we can access the randomly
  # generated shipping/tracking IDs later, since we're not using actual
  # information from a product shipping database.
  shipment_req1 = Google::Apis::ContentV2::OrdersShipLineItemsRequest.new
  shipment_req1.operation_id = new_operation_id()
  shipment_req1.line_items = [ line_item ]
  shipment_req1.carrier =
      current_order.line_items[0].shipping_details.method_prop.carrier
  shipment_req1.shipment_id = "#{Random.rand(2**32 - 1)}"
  shipment_req1.tracking_id = "#{Random.rand(2**32 - 1)}"
  resp = content_api.shiplineitems_order(merchant_id, order_id, shipment_req1)
  puts "done (#{resp.execution_status})."
  puts

  current_order = content_api.get_order(merchant_id, order_id)
  print_order(current_order)
  puts

  # Now we ship the rest.
  print "Notifying Google about shipment of second line item... "
  line_item = Google::Apis::ContentV2::OrderShipmentLineItemShipment.new
  line_item.line_item_id = current_order.line_items[1].id
  line_item.quantity = current_order.line_items[1].quantity_pending
  shipment_req2 = Google::Apis::ContentV2::OrdersShipLineItemsRequest.new
  shipment_req2.operation_id = new_operation_id()
  shipment_req2.line_items = [ line_item ]
  shipment_req2.carrier =
      current_order.line_items[1].shipping_details.method_prop.carrier
  shipment_req2.shipment_id = "#{Random.rand(2**32 - 1)}"
  shipment_req2.tracking_id = "#{Random.rand(2**32 - 1)}"
  resp = content_api.shiplineitems_order(merchant_id, order_id, shipment_req2)
  puts "done (#{resp.execution_status})."
  puts

  current_order = content_api.get_order(merchant_id, order_id)
  print_order(current_order)
  puts

  # First item arrives to the customer.
  print "Notifying Google about delivery of first line item... "
  req = Google::Apis::ContentV2::OrdersUpdateShipmentRequest.new
  req.operation_id = new_operation_id()
  req.carrier = shipment_req1.carrier
  req.tracking_id = shipment_req1.tracking_id
  req.shipment_id = shipment_req1.shipment_id
  req.status = "delivered"
  resp = content_api.update_order_shipment(merchant_id, order_id, req)
  puts "done (#{resp.execution_status})."
  puts

  current_order = content_api.get_order(merchant_id, order_id)
  print_order(current_order)
  puts

  # Second item arrives to the customer.
  print "Notifying Google about delivery of second line item... "
  req = Google::Apis::ContentV2::OrdersUpdateShipmentRequest.new
  req.operation_id = new_operation_id()
  req.carrier = shipment_req2.carrier
  req.tracking_id = shipment_req2.tracking_id
  req.shipment_id = shipment_req2.shipment_id
  req.status = "delivered"
  resp = content_api.update_order_shipment(merchant_id, order_id, req)
  puts "done (#{resp.execution_status})."
  puts

  current_order = content_api.get_order(merchant_id, order_id)
  print_order(current_order)
  puts

  # Customer returns one of first item since it was broken on arrival.
  print "Notifying Google about return of first line item... "
  req = Google::Apis::ContentV2::OrdersReturnLineItemRequest.new
  req.operation_id = new_operation_id()
  req.line_item_id = current_order.line_items[0].id
  req.quantity = 1
  req.reason = "productArrivedDamaged"
  req.reason_text = "Item broken upon receipt."
  resp = content_api.return_order_line_item(merchant_id, order_id, req)
  puts "done (#{resp.execution_status})."
  puts

  current_order = content_api.get_order(merchant_id, order_id)
  print_order(current_order)
  puts
end

# All (non-test) requests that change an order must have a unique operation
# ID over the lifetime of the order to enable Google to detect and reject
# duplicate requests.
# Here, we just use a nonce and bump it each time, since we're not retrying.
# This isn't thread-safe, but okay here as we're sending the requests
# sequentially.
@nonce = 0
def new_operation_id()
  operation_id = "#{@nonce}"
  @nonce += 1
  return operation_id
end

def list_all_orders(content_api, merchant_id, **args)
  content_api.list_orders(merchant_id, **args) do |res, err|
    if err
      handle_errors(err)
      exit
    end

    unless res.resources
      puts 'No results.'
      return
    end

    res.resources.each { |order| print_order(order) }

    return unless res.next_page_token
    args[:page_token] = res.next_page_token
    list_all_orders(content_api, merchant_id, **args)
  end

end

if __FILE__ == $0
  options = ArgParser.parse(ARGV)
  config, content_api = service_setup(options, use_sandbox = true)
  order_workflow(content_api, config.merchant_id)
end

File: java/src/main/java/shopping/content/v2_1/samples/orders/OrdersWorkflow.java
package shopping.content.v2_1.samples.orders;

import static shopping.common.BaseOption.NO_CONFIG;

import com.google.api.services.content.ShoppingContent;
import com.google.api.services.content.model.Order;
import com.google.api.services.content.model.OrderLineItem;
import com.google.api.services.content.model.OrderShipmentLineItemShipment;
import com.google.api.services.content.model.OrdersAcknowledgeRequest;
import com.google.api.services.content.model.OrdersCancelLineItemRequest;
import com.google.api.services.content.model.OrdersCreateTestOrderRequest;
import com.google.api.services.content.model.OrdersCustomBatchRequestEntryShipLineItemsShipmentInfo;
import com.google.api.services.content.model.OrdersListResponse;
import com.google.api.services.content.model.OrdersReturnRefundLineItemRequest;
import com.google.api.services.content.model.OrdersShipLineItemsRequest;
import com.google.api.services.content.model.OrdersUpdateMerchantOrderIdRequest;
import com.google.api.services.content.model.OrdersUpdateShipmentRequest;
import com.google.common.collect.ImmutableList;
import java.io.File;
import java.io.IOException;
import java.util.Random;
import org.apache.commons.cli.CommandLine;
import shopping.common.BaseOption;
import shopping.content.v2_1.samples.ContentConfig;
import shopping.content.v2_1.samples.ContentWorkflowSample;

/**
 * Sample that runs through an entire test order workflow. We run this sample on the sandbox API
 * endpoint, so that we have access to test order creation and don't accidentally mutate real
 * orders.
 */
public class OrdersWorkflow extends ContentWorkflowSample {
  private int nonce = 0;
  private final Random random = new Random();

  private OrdersWorkflow(ShoppingContent content, ShoppingContent sandbox, ContentConfig config) {
    super(content, sandbox, config);
  }

  @Override
  public void execute() throws IOException {
    checkNonMCA();

    // Create a new test order using the template1 template. Normally orders would be
    // automatically populated by Google in the non-sandbox version, and we'd skip
    // to finding out what orders are currently waiting for us.
    System.out.print("Creating test order... ");
    String orderId =
        sandbox
            .orders()
            .createtestorder(
                config.getMerchantId(),
                new OrdersCreateTestOrderRequest().setTemplateName("template1"))
            .execute()
            .getOrderId();
    System.out.println("done.");
    System.out.printf("Order \"%s\" created.%n", orderId);
    System.out.println();

    // List the unacknowledged orders.
    System.out.printf("Listing unacknowledged orders for merchant %d:%n", config.getMerchantId());
    ShoppingContent.Orders.List listCall =
        sandbox.orders().list(config.getMerchantId()).setAcknowledged(false);
    do {
      OrdersListResponse page = listCall.execute();
      for (Order product : page.getResources()) {
        OrdersUtils.printOrder(product);
      }
      if (page.getNextPageToken() == null) {
        break;
      }
      listCall.setPageToken(page.getNextPageToken());
    } while (true);
    System.out.println();

    // Acknowledge the newly received order. (Normally we'd do this for the orders returned by the
    // list call, but here, we'll just use the one we got from creating the test order.)
    System.out.printf("Acknowledging order \"%s\"... ", orderId);
    System.out.printf(
        "done with status \"%s\".%n",
        sandbox
            .orders()
            .acknowledge(
                config.getMerchantId(),
                orderId,
                new OrdersAcknowledgeRequest().setOperationId(newOperationId()))
            .execute()
            .getExecutionStatus());
    System.out.println();

    // Set the new order's merchant order ID.
    String merchantOrderId = "test order " + String.valueOf(random.nextLong());
    System.out.printf("Updating merchant order ID to \"%s\"... ", merchantOrderId);
    System.out.printf(
        "done with status \"%s\".%n",
        sandbox
            .orders()
            .updatemerchantorderid(
                config.getMerchantId(),
                orderId,
                new OrdersUpdateMerchantOrderIdRequest()
                    .setMerchantOrderId(merchantOrderId)
                    .setOperationId(newOperationId()))
            .execute()
            .getExecutionStatus());
    System.out.println();

    Order currentOrder = getOrderByMerchantOrderId(merchantOrderId);
    OrdersUtils.printOrder(currentOrder);
    System.out.println();

    // Oops, not enough stock for all the Chromecasts, so we cancel one of them.
    System.out.print("Canceling one Chromecast order... ");
    System.out.printf(
        "done with status \"%s\".%n",
        sandbox
            .orders()
            .cancellineitem(
                config.getMerchantId(),
                orderId,
                new OrdersCancelLineItemRequest()
                    .setOperationId(newOperationId())
                    .setLineItemId(currentOrder.getLineItems().get(0).getId())
                    .setQuantity(1L)
                    .setReason("noInventory")
                    .setReasonText("Ran out of inventory while fulfilling request."))
            .execute()
            .getExecutionStatus());
    System.out.println();

    currentOrder = getOrder(orderId);
    OrdersUtils.printOrder(currentOrder);
    System.out.println();

    // Advance the test order to the shippable state.  Normally this would be done by
    // Google when an order is no longer cancelable by the customer, but here we need
    // to do it ourselves.
    System.out.print("Advancing test order... ");
    sandbox.orders().advancetestorder(config.getMerchantId(), orderId).execute();
    System.out.println("done.");
    System.out.println();

    currentOrder = getOrder(orderId);
    OrdersUtils.printOrder(currentOrder);
    System.out.println();

    // To simulate partial fulfillment, we'll pick the first line item from our test order
    // and ship the amount that's still pending.
    System.out.print("Notifying Google about shipment of first line item... ");
    OrderLineItem item1 = currentOrder.getLineItems().get(0);
    // Storing this in a variable so we can access the shipment/tracking IDs later.
    OrdersShipLineItemsRequest shipReq1 =
        new OrdersShipLineItemsRequest()
            .setOperationId(newOperationId())
            .setLineItems(
                ImmutableList.of(
                    new OrderShipmentLineItemShipment()
                        .setLineItemId(item1.getId())
                        .setQuantity(item1.getQuantityPending())))
                .setShipmentInfos(
                        ImmutableList.of(
                                new OrdersCustomBatchRequestEntryShipLineItemsShipmentInfo()
                                        .setCarrier(item1.getShippingDetails().getMethod()
                                                        .getCarrier())
                                        .setShipmentId(String.valueOf(random.nextLong()))
                                        .setTrackingId(String.valueOf(random.nextLong()))
                        ));
    System.out.printf(
        "done with status \"%s\".%n",
        sandbox
            .orders()
            .shiplineitems(config.getMerchantId(), orderId, shipReq1)
            .execute()
            .getExecutionStatus());
    System.out.println();

    currentOrder = getOrder(orderId);
    OrdersUtils.printOrder(currentOrder);
    System.out.println();

    // Now we ship the rest.
    System.out.print("Notifying Google about shipment of second line item... ");
    OrderLineItem item2 = currentOrder.getLineItems().get(1);
    OrdersShipLineItemsRequest shipReq2 =
        new OrdersShipLineItemsRequest()
            .setOperationId(newOperationId())
            .setLineItems(
                ImmutableList.of(
                    new OrderShipmentLineItemShipment()
                        .setLineItemId(item2.getId())
                        .setQuantity(item2.getQuantityPending())))
            .setShipmentInfos(
                    ImmutableList.of(
                      new OrdersCustomBatchRequestEntryShipLineItemsShipmentInfo()
                              .setCarrier(item2.getShippingDetails().getMethod().getCarrier())
                              .setShipmentId(String.valueOf(random.nextLong()))
                              .setTrackingId(String.valueOf(random.nextLong()))
            ));
    System.out.printf(
        "done with status \"%s\".%n",
        sandbox
            .orders()
            .shiplineitems(config.getMerchantId(), orderId, shipReq2)
            .execute()
            .getExecutionStatus());
    System.out.println();

    currentOrder = getOrder(orderId);
    OrdersUtils.printOrder(currentOrder);
    System.out.println();

    // First item arrives to the customer.
    System.out.print("Notifying Google about delivery of first line item... ");
    OrdersCustomBatchRequestEntryShipLineItemsShipmentInfo l2 = shipReq1.getShipmentInfos().get(0);
    System.out.printf(
        "done with status \"%s\".%n",
        sandbox
            .orders()
            .updateshipment(
                config.getMerchantId(),
                orderId,
                new OrdersUpdateShipmentRequest()
                    .setOperationId(newOperationId())
                    .setCarrier(shipReq1.getShipmentInfos().get(0).getCarrier())
                    .setTrackingId(shipReq1.getShipmentInfos().get(0).getTrackingId())
                    .setShipmentId(shipReq1.getShipmentInfos().get(0).getShipmentId())
                    .setStatus("delivered"))
            .execute()
            .getExecutionStatus());
    System.out.println();

    currentOrder = getOrder(orderId);
    OrdersUtils.printOrder(currentOrder);
    System.out.println();

    // Second item arrives.
    System.out.print("Notifying Google about delivery of second line item... ");
    System.out.printf(
        "done with status \"%s\".%n",
        sandbox
            .orders()
            .updateshipment(
                config.getMerchantId(),
                orderId,
                new OrdersUpdateShipmentRequest()
                    .setOperationId(newOperationId())
                        .setCarrier(shipReq2.getShipmentInfos().get(0).getCarrier())
                        .setTrackingId(shipReq2.getShipmentInfos().get(0).getTrackingId())
                        .setShipmentId(shipReq2.getShipmentInfos().get(0).getShipmentId())
                        .setStatus("delivered"))
            .execute()
            .getExecutionStatus());
    System.out.println();

    currentOrder = getOrder(orderId);
    OrdersUtils.printOrder(currentOrder);
    System.out.println();

    // Customer returns the first delivered item because it's malfunctioning.
    System.out.print("Notifying Google about return of first line item... ");
    System.out.printf(
        "... done with status \"%s\".%n",
        sandbox
            .orders()
            .returnrefundlineitem(
                config.getMerchantId(),
                orderId,
                new OrdersReturnRefundLineItemRequest()
                    .setOperationId(newOperationId())
                    .setLineItemId(item1.getId())
                    .setQuantity(1L)
                    .setReason("productArrivedDamaged")
                    .setReasonText("Item malfunctioning upon receipt."))
            .execute()
            .getExecutionStatus());
    System.out.println();

    currentOrder = getOrder(orderId);
    OrdersUtils.printOrder(currentOrder);
    System.out.println();
  }

  private String newOperationId() {
    String ret = String.valueOf(nonce);
    nonce++;
    return ret;
  }

  private Order getOrder(String orderId) throws IOException {
    System.out.printf("Retrieving order \"%s\"... ", orderId);
    Order ret = sandbox.orders().get(config.getMerchantId(), orderId).execute();
    System.out.println("done.");
    System.out.println();
    return ret;
  }

  private Order getOrderByMerchantOrderId(String merchantOrderId) throws IOException {
    System.out.printf("Retrieving merchant order \"%s\"... ", merchantOrderId);
    Order ret =
        sandbox
            .orders()
            .getbymerchantorderid(config.getMerchantId(), merchantOrderId)
            .execute()
            .getOrder();
    System.out.println("done.");
    System.out.println();
    return ret;
  }

  public static void main(String[] args) throws IOException {
    CommandLine parsedArgs = BaseOption.parseOptions(args);
    File configPath = null;
    if (!NO_CONFIG.isSet(parsedArgs)) {
      configPath = BaseOption.checkedConfigPath(parsedArgs);
    }
    ContentConfig config = ContentConfig.load(configPath);

    ShoppingContent.Builder builder = createStandardBuilder(parsedArgs, config);
    ShoppingContent content = createService(builder);
    ShoppingContent sandbox = createSandboxContentService(builder);
    retrieveConfiguration(content, config);

    new OrdersWorkflow(content, sandbox, config).execute();
  }
}
